hilti/toolchain/include/base/optional-ref.h

```diff
@@ -0,0 +1,83 @@
+// From https://github.com/Chlorie/clu/tree/4d8bc5cfa780b9ba1d19d2e911c5bb80246a6fa4. Extended.
+
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <utility>
+
+namespace hilti {
+template<typename T>
+class optional_ref final {
+private:
+    T* ptr_ = nullptr;
+
+public:
+    using nonConstT = typename std::remove_const<T>::type;
+    constexpr optional_ref() noexcept = default;
+    constexpr optional_ref(std::nullopt_t) noexcept {}
+    constexpr optional_ref(std::nullptr_t) noexcept {}
+    constexpr optional_ref(T& reference) noexcept : ptr_(std::addressof(reference)) {}
```

Comment:
Not sure on my rule of X, but does this allow constructing from r-value refs? If it does it would be great to delete that constructor to make it harder to construct dangling optional refs.

Reply:
ack, will do